test/library/packages/Shlex/Shlex.chpl

```diff
@@ -0,0 +1,442 @@
+/*
+ * Copyright 2020-2022 Hewlett Packard Enterprise Development LP
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+  A lexical anayzer class for simple shell-like syntaxes.
+  Implementation and design inspired by Python's Shlex module: 
+  <https://docs.python.org/3/library/shlex.html>
+  This module provides support for split, join and quote operations.
+*/
+
+module Shlex {
+  private use List;
+  private use Regex;
+  class shlex{
+    var instream: string;
+    var infile: string;
+    var posix: bool;
+    var lineno: int;
+    param debug: int;
+    var commenters: string;
+    var wordchars: string;
+    var whitespace: string;
+    var whitespace_split: bool;
+    var quotes: string;
+    var escapse: string;
+    var escapsedquotes: string;
+    var state: string;
+    var token: string;
+    var tokindex: int;
+    var punctuation_chars: string;
+    var pushback: list(string);
+    var _pushback_chars: list(string);
+    var custom_punctuation_chars: string;
+
+    /* 
+      Initializes the required variables of class `shlex`.
+      Used in new expressions.
+
+      :arg instream: The string to be parsed by shlex.
+      :type instream: `string`
+      :arg posix: Argument to decide parsing rules (POSIX or non-POSIX).
+                  Defaults to `false`.
+      :type posix: `bool`
+      :arg punctuation_chars: If set to true, contiguous runs of punctuation
+                              characters will be returned as a single token.
+                              Defaults to `false`. 
+      :type punctuation_chars: `bool`
+      :arg commenters: Strings of characters that are recognized as comment
+                        beginners. All characters from comment beginners to end
+                        of line are ignored. Defaults to `#`.
+      :type commenters: `string`
+      :arg custom_punctuation_chars: All the character present in this string
+                                      will be considered as punctuation
+                                      characters. NOTE: This field is only used
+                                      if the argument `punctuation_chars` is
+                                      set to true. Defaults to `();|&`.
+      :type custom_punctuation_chars: `string`
+    */
+    proc init(instream:string, posix: bool = false,
```

Comment:
I have a feeling your chpldoc error is coming from somewhere in the docstring above. Does this initializer need to be publicly documented? Will users ever initialize their own shlex class? I assumed they would use the standalone functions.

Reply:
Some of the conditions can not be initialized by ``split`` function, such as having ``punctuation_chars`` set to true and ``whitespace_split`` set to true simultaneously, or both of them to false. One of the tests present in ``ParsingEmptyStrings.chpl`` have such a requirement where ``split`` function does not satisfy. Also, other requirements such as changing ``debug`` or ``wordchars`` could be done easily if an user is allowed to use ``init`` function. If we have to make the standalone function ``split()`` to handle these requirements, its argument list would also be same as the variables in ``shlex`` class. Please confirm if ``init`` function should not be user-facing, or is it better to document it and allow the users to configure necessary variables. Also, the Python shlex implementation has less user flexibility provided by ``split`` function, and most of its test uses constructor for configuring  variables.